### PR TITLE
Attention mask for Flux & SD3

### LINKED ISSUE #8673: Attention masks are missing in SD3 to mask out text padding tokens
### Describe the bug

In the attention implementation of SD3, attention masks currently are not used. This will result in inconsistent outputs for the different values `max_seq_length` where padding exists in text tokens as the attention scores of padding tokens are non-zero. This issue has been discussed in https://github.com/huggingface/diffusers/discussions/8628, and is created to track the progress of fixing this problem.

Thanks @sayakpaul for the discussion.

### Reproduction

n/a

### Logs

_No response_

### System Info

n/a

### Who can help?

_No response_

### LINKED ISSUE #10025: attention mask for transformer Flux
### Describe the bug

Is it possible to get back the `attention_mask` argument in the flux attention processor 

```
hidden_states = F.scaled_dot_product_attention(query, key, value, dropout_p=0.0, is_causal=False,attn_mask=attention_mask)
```

https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/attention_processor.py#L1910

in order to tweak things a bit ? otherwise the argument  `attention_mask`  is unused.

Thanks a lot 

### Reproduction

pip install diffusers

### Logs

_No response_

### System Info

Ubuntu

### Who can help?

@yiyixuxu @sayakpaul @DN6 @asomoza

### PR DESCRIPTION
**Pull request by rootonchair**
# What does this PR do?

Fixes #10025
Fixes #8673

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by rootonchair** [comment]
@sayakpaul @yiyixuxu, how should I test this feature? modify the original flux pipeline?

**Comment by bghira** [comment]
refer to [this](https://github.com/bghira/SimpleTuner/blob/main/helpers/models/flux/transformer.py) flux transformer implementation for attn masking details

File: src/diffusers/models/attention_processor.py [hunk 1]
```diff
@@ -1129,11 +1129,27 @@ def __call__(
...
 
+            query_pad_size = query.size(2)
+            key_pad_size = key.size(2)
+
             query = torch.cat([query, encoder_hidden_states_query_proj], dim=2)
             key = torch.cat([key, encoder_hidden_states_key_proj], dim=2)
             value = torch.cat([value, encoder_hidden_states_value_proj], dim=2)
 
-        hidden_states = F.scaled_dot_product_attention(query, key, value, dropout_p=0.0, is_causal=False)
+            if attention_mask is not None:
```
Comment (yiyixuxu) [inline src/diffusers/models/attention_processor.py:1139]: well I think what we want is not to have a specific implementation to apply attention_mask for flux, is just to allow it to pass down all the way from pipeline, to transformer and the to attention processor so user can experiment with a custom attention mask 

cc @christopher5106 is what I described here something you have in mind?

Comment (yiyixuxu) [reply src/diffusers/models/attention_processor.py:1139]: and if you do have an specific implementation that we want to add to diffusers,  maybe you can run some experiments to help us decide if it's meaningful

Comment (bghira) [reply src/diffusers/models/attention_processor.py:1139]: it's the encoder attention mask though just following convention of pixart and other DiT that rely on attention masking. masking the attention arbitrarily doesn't unlock new use cases, does it? if so, providing examples of those would be nice.

Comment (yiyixuxu) [reply src/diffusers/models/attention_processor.py:1139]: @rootonchair yes feel free to modify pipeline/model to test, and provide the experiments results to us:)
>how should I test this feature? modify the original flux pipeline?

Comment (yiyixuxu) [reply src/diffusers/models/attention_processor.py:1139]: cc @christopher5106 - would you be able to provide a use case? since it was the original ask
> masking the attention arbitrarily doesn't unlock new use cases, does it? if so, providing examples of those would be nice.

**Comment by rootonchair** [comment]
> refer to [this](https://github.com/bghira/SimpleTuner/blob/main/helpers/models/flux/transformer.py) flux transformer implementation for attn masking details

This implementation is a little confuse for me. If the encoder attention mask is `[1,1,1,0,0]`
Should the attention mask of the joint attention process be (assume the sequence length of images is 0)
```
[
[1,1,1,0,0],
[1,1,1,0,0],
[1,1,1,0,0],
[0,0,0,0,0],
[0,0,0,0,0]
]
```
Simply broadcast the mask would result in 

```
[
[1,1,1,0,0],
[1,1,1,0,0],
[1,1,1,0,0],
[1,1,1,0,0],
[1,1,1,0,0]
]
```

while the two last token are just padding token
```

```
[
[1,1,1,0,0],
[1,1,1,0,0],
[1,1,1,0,0],
[1,1,1,0,0], <--- query's pad token
[1,1,1,0,0] <--- query's pad token
]
```
```

**Comment by bghira** [comment]
no flux doesnt mask padding tokens. doing so harms prompt adherence

File: src/diffusers/models/attention_processor.py (hunk 1 above)
Comment (bghira) [reply src/diffusers/models/attention_processor.py:1139]: checking the softmax scores for padded positions.

**Comment by rootonchair** [comment]
This is my result testing with Flux pipeline
```python
import torch
from diffusers import FluxPipeline

pipe = FluxPipeline.from_pretrained("black-forest-labs/FLUX.1-dev", torch_dtype=torch.bfloat16)
pipe.enable_sequential_cpu_offload()
pipe.vae.enable_slicing()
pipe.vae.enable_tiling()
pipe.to(torch.float16)


prompt = [
    "a tiny astronaut hatching from an egg on the moon", 
    "A cat holding a sign that says hello world"
]
attention_mask = pipe.tokenizer_2(
    prompt,
    padding="max_length",
    max_length=512,
    truncation=True,
    return_length=False,
    return_overflowing_tokens=False,
    return_tensors="pt",
).attention_mask
attention_mask = attention_mask.to(device="cuda", dtype=torch.float16)
print(attention_mask)
print(attention_mask.argmin(axis=1))
out = pipe(
    prompt=prompt,
    guidance_scale=3.5,
    height=768,
    width=1360,
    num_inference_steps=50,
    joint_attention_kwargs = {'attention_mask': attention_mask},
    generator=torch.Generator(device="cuda").manual_seed(42)
).images
out[0].save("image.png")
out[1].save("image1.png")
```

Without mask
old_image
old_image1

With mask
new_image
new_image1

**Comment by bghira** [comment]
patch embed artifacts in the masked one.

**Comment by rootonchair** [comment]
For SD3, it doesn't have the same effect. I will try to figure out why and come up with an test script soon

**Comment by rootonchair** [comment]
For SD3 I use the below script:
```python
import torch
from diffusers import StableDiffusion3Pipeline

pipe = StableDiffusion3Pipeline.from_pretrained(
    "stabilityai/stable-diffusion-3-medium-diffusers", torch_dtype=torch.float16
)
pipe.enable_model_cpu_offload()

prompt = [
    "", # negative prompt for first image
    "", # negative prompt fort second image
    "smiling cartoon dog sits at a table, coffee mug on hand, as a room goes up in flames. “This is fine,” the dog assures himself.", 
    "A cat holding a sign that says hello world",
]
t5_attention_mask = pipe.tokenizer_3(
    prompt,
    padding="max_length",
    max_length=256,
    truncation=True,
    add_special_tokens=True,
    return_tensors="pt",
).attention_mask

attention_mask = t5_attention_mask 
attention_mask = attention_mask.to(device="cuda", dtype=torch.float16)

prompt = prompt[2:]
print(prompt)
image = pipe(
    prompt,
    joint_attention_kwargs={'attention_mask': attention_mask},
    generator = torch.Generator(device="cuda").manual_seed(42),
).images
image[0].save("new_sd_image.png")
image[1].save("new_sd_image1.png")
```

without mask
old_sd_image


old_sd_image1


with mask
new_sd_image

new_sd_image1

**Comment by christopher5106** [comment]
it looks like there is a bug

**Comment by bghira** [comment]
yeah i'm not too sure what's happening here anymore. the attention_mask is ignoring image token inputs? it is just for text encoder attn mask now?

i thought the idea was to add encoder_attention_mask and image_attention_mask parameters and make them cooperate with joint_attention_kwargs attention_mask which would supercede those two.

the test should not change the outputs at all, and yours does, which indicate there is something wrong.

**Review by joangava — APPROVED** [review]

**Review by joangava — APPROVED** [review]

**Comment by IorHon** [comment]
Has anyone already gotten it working properly? I'm willing to pay for the work to anyone who can integrate it into Comfyui for me.

**Review by sins921 — APPROVED** [review]

**Review by joangava — APPROVED** [review]

**Comment by Priyachakraborty-hash** [comment]
Awesome!

**Review by nexoracontrol-ops — APPROVED** [review]
